### PR TITLE
Fix: Prevent download of inactive ws zip (without 500)

### PR DESCRIPTION
**Pull request by MarcelGeo**
Fixes https://github.com/MerginMaps/server-private/issues/2381

- prevent 500 error code for case where user is_anonymous and stayed on project tree page and trying to donwload file of inactive ws
- admin is still available to download such project

Error message now (workspace is not active, does not exists for users anymore):
image

### CODE REVIEW

File: server/mergin/sync/permissions.py [hunk 1]
```diff
@@ -157,12 +157,24 @@ def get_user_project_role(
             return ProjectRole.READER
         return None
 
+def is_inactive_workspace(workspace):
```
Comment (harminius) [inline server/mergin/sync/permissions.py:160]: I'd prefer a positive funciton name: `is_active_workspace`

File: server/mergin/sync/permissions.py [hunk 2]
```diff
@@ -157,12 +157,24 @@ def get_user_project_role(
...
+        workspace (Workspace): The workspace to check.
+
+    Returns:
+        bool: True if the workspace is not active and the current user is not an admin or is anonymous, False otherwise.
+    """
+    is_admin = not current_user.is_anonymous and current_user.is_admin
     if not workspace:
         abort(404)
-    if not workspace.is_active and not current_user.is_admin:
+    return not workspace.is_active and (not is_admin or is_anonymous)
```
Comment (harminius) [inline server/mergin/sync/permissions.py:173]: `is_anonymous` -> `current_user.is_anonymous`?

**Comment by harminius** [comment]
It is worth adding some tests.

**Comment by MarcelGeo** [comment]
> It is worth adding some tests.

Sure. It is hard to simulate AnonymouUser in tests ..

**Comment by MarcelGeo** [comment]
Tests added in private repo :brain:

File: server/mergin/sync/permissions.py [hunk 3]
```diff
@@ -158,11 +158,25 @@ def get_user_project_role(
...
-    workspace = current_app.ws_handler.get_by_name(ws)
+def is_active_workspace(workspace):
+    """
+    Checks if the given workspace is active and if the current user is an admin or anonymous.
+
+    Args:
+        workspace (Workspace): The workspace to check.
+
+    Returns:
+        bool: True if the workspace is not active and the current user is not an admin or is anonymous, False otherwise.
```
Comment (harminius) [inline server/mergin/sync/permissions.py:169]: Update description, please

**Review by harminius — APPROVED** [review]
Looks good.

File: server/mergin/sync/permissions.py [hunk 4]
```diff
@@ -158,11 +158,25 @@ def get_user_project_role(
...
+    """
     if not workspace:
         abort(404)
```
Comment (varmar05) [inline server/mergin/sync/permissions.py:172]: this is not bool, also we do not need such check, I assume we won't pass None in arg

File: server/mergin/sync/permissions.py [hunk 5]
```diff
@@ -158,11 +158,25 @@ def get_user_project_role(
         return None
 
 
-def require_project(ws, project_name, permission):
-    workspace = current_app.ws_handler.get_by_name(ws)
+def is_active_workspace(workspace):
```
Comment (varmar05) [inline server/mergin/sync/permissions.py:161]: do we need this function at all? there is already similar check https://github.com/MerginMaps/server/blob/gh%232381-prevent-download-anonymous-inactive-ws/server/mergin/sync/workspace.py#L68

we would only need to handle current_user properly

Comment (MarcelGeo) [reply server/mergin/sync/permissions.py:161]: I am not sure, this just abstracted code from require_project and required_project_by_uuid .

File: server/mergin/sync/permissions.py (hunk 4 above)
Comment (MarcelGeo) [reply server/mergin/sync/permissions.py]: https://github.com/MerginMaps/server/pull/239/commits/b267f71681394711a0d6c30b38794efdba8a374c

File: server/mergin/sync/permissions.py (hunk 3 above)
Comment (MarcelGeo) [reply server/mergin/sync/permissions.py]: https://github.com/MerginMaps/server/pull/239/commits/b267f71681394711a0d6c30b38794efdba8a374c

File: server/mergin/sync/permissions.py (hunk 5 above)
Comment (varmar05) [reply server/mergin/sync/permissions.py:161]: I am only worried we use various similar functions for doing the same thing and most probably not used everywhere - we might want to revisit this at some point.

**Review by varmar05 — APPROVED** [review]

File: server/mergin/sync/permissions.py (hunk 5 above)
Comment (MarcelGeo) [reply server/mergin/sync/permissions.py:161]: Thanks :) But currently I am no table to do some changes there .

**Pull request merged**